checks.go

```diff
@@ -0,0 +1,203 @@
+// checks.go contains checks that are identical for both Linux and Windows.
+
+package main
+
+import (
+	"crypto/sha256"
+	"encoding/hex"
+	"errors"
+	"fmt"
+	"os"
+	"path/filepath"
+	"reflect"
+	"regexp"
+	"strings"
+)
+
+// check is the smallest unit that can show up on a scoring report. It holds all
+// the conditions for a check, and its message and points (autogenerated or
+// otherwise).
+type check struct {
+	Message      string
+	Points       int
+	Fail         []cond
+	Pass         []cond
+	PassOverride []cond
+}
+
+// cond, or condition, is the parameters for a given test within a check.
+type cond struct {
+	Type    string
+	Path    string
+	Hash    string
+	Cmd     string
+	User    string
+	Group   string
+	Program string
+	Version string
+	Name    string
+	Key     string
+	Value   string
+	After   string
+}
+
+func (c cond) String() string {
+	output := ""
+	v := reflect.ValueOf(c)
+	typeOfS := v.Type()
+
+	for i := 0; i < v.NumField(); i++ {
+		if v.Field(i).String() == "" {
+			continue
+		}
+		output += fmt.Sprintf("\t%s: %v\n", typeOfS.Field(i).Name, v.Field(i).String())
+	}
+	return output
+}
+
+func handleReflectPanic(condFunc string) {
+	if r := recover(); r != nil {
+		fail("Check type does not exist: "+condFunc, "("+fmt.Sprint(r)+")")
+	}
+}
+
+// runCheck executes a single condition check.
+func runCheck(cond cond) bool {
+	if err := deobfuscateCond(&cond); err != nil {
+		fail(err.Error())
+	}
+	defer obfuscateCond(&cond)
+	debug("Running condition:\n", cond)
+
+	condFunc := ""
+	negation := false
+	condEnding := cond.Type[len(cond.Type)-3 : len(cond.Type)]
+	if condEnding == "Not" {
+		negation = true
+		condFunc = cond.Type[:len(cond.Type)-3]
+	} else {
+		condFunc = cond.Type
+	}
+
+	// Catch panic if check type doesn't exist
+	defer handleReflectPanic(condFunc)
+
+	// Using reflection to find the correct function to call.
+	vals := reflect.ValueOf(cond).MethodByName(condFunc).Call([]reflect.Value{})
+	result := vals[0].Bool()
+	err := vals[1]
+	if negation {
+		return err.IsNil() && !result
+	}
+	debug("Result is", result, "and error is", err)
+	return err.IsNil() && result
+}
+
+// CommandContains checks if a given shell command contains a certain output.
+// This check will always fail if the command returns an error.
```

Comment:
Why is `CommandContains` and `CommandOutput` here, but `Command` (the make sure it runs without error) itself is alone in `linux.go`? Especially considering I don't see a similar check in `windows.go` (I checked only the top because it seemed alphabetized)

Reply:
Good question, I removed `Command` check for Windows because it was awful and not reliable at all for checking if a command succeeded. For Windows, I think we should encourage using the API or at least Command{Contains,Output}.